File: src/core/schema/tools/union.rb
require 'core/schema/code/factory'

#=begin

#This file creates the copy of two structures.
#The copy U of structures A and B has all the
#objects of both A and B. 

#=end
require 'enso'

module Union
  
  class CopyInto
    def initialize(factory)
      @memo = {}
      @factory = factory
    end
  
    # computes the copy of structures given root nodes a and b  
    def copy(a, b)
      build(a, b)
      link(true, a, b)
    end
    
    # build walks the spine of the two structures and matches up
    # corresponding objects. The most important thing to keep in 
    # mind is that either a or b (or both) can be nil, if there is no
    # corresponding structure in the other structure.
    # This function builds all the new objects and also initialized
    # the primitive fields. Primitive fields must be initialized
    # first so that the keys will be defined before objects are added
    # to keyed collections
    def build(a, b)
      if !a.nil?
        raise "Union of incompatible objects #{a} and #{b}" if a && b && a.schema_class.name != b.schema_class.name
        @memo[a.identity] = new = b || @factory[a.schema_class.name]
        #puts "BUILD #{a} + #{b} ==> #{new}"
        new.schema_class.fields.each do |field|
          a_val = begin a[field.name]
          rescue 
            nil 
          end
          b_val = begin b[field.name] 
          rescue 
            nil 
          end
          if !a_val.nil? or !b_val.nil?
            if field.type.is_a?("Primitive")
              if !a_val.nil?
                if a && b && a_val != b_val then
                  puts "UNION WARNING: changing #{new}.#{field.name} from '#{b_val}' to '#{a_val}'"
                end
                new[field.name] = a_val
              end
            elsif field.traversal
              if !field.many
                build(a_val, b_val)
              else
                a_val.each_with_match(b_val) do |a_item, b_item|
                  build(a_item, b_item)
                end
              end
            end
          else
            if !new[field.name].nil?
              puts "skipping #{new}.#{field.name} as #{new[field.name]}"
            end
          end
        end
      end
    end
  
    # creates the cross-links in the CopyInto. The "traversal" field is used
    # to go one stage past the spine, to relate linked objects.
    def link(traversal, a, b)
      if a.nil?
        b 
      else
        new = @memo[a.identity]
        #puts "LINK #{a} + #{b} ==> #{new}"
        if !new
          p @memo
          raise "Traversal did not visit every object a=#{a} b=#{b}" 
        end
        if traversal
          a.schema_class.fields.each do |field|
            a_val = a[field.name]
            b_val = b && b[field.name]
            if !field.type.is_a?("Primitive")
              if !field.many
                val = link(field.traversal, a_val, b_val)
                new[field.name] = val
              else
                a_val.each_with_match(b_val) do |a_item, b_item|
                  item = link(field.traversal, a_item, b_item)
                  new[field.name] << item  # TODO: WHY WAS THIS HERE: unless new[field.name].include? item
                end
              end
            end
          end
        end
        new
      end
    end
  end        
  
  def self.CopyInto(factory, a, b)
    CopyInto.new(factory).copy(a, b)
  end   
  
  def self.Copy(factory, a)
    CopyInto.new(factory).copy(a, nil).finalize
  end
  
  def self.Clone(a)
    Copy(a.factory, a)
  end
      
  def self.Union(factory, *parts)
    copier = CopyInto.new(factory)
    result = nil
    parts.each do |part|
      result = copier.copy(part, result)
    end
    result.finalize
  end   
  
  def self.union(a, b)
    f = Factory::SchemaFactory.new(a.graph_identity.schema)
    Union(f, a, b)
  end
end

File: src/js/core/schema/tools/union.js
'use strict'

//// Union ////

var cwd = process.cwd() + '/';
var Factory = require(cwd + "core/schema/code/factory.js");
var Enso = require(cwd + "enso.js");

var Union;

var Copy = function(factory, a) {
  var self = this;
  return CopyInto.new(factory).copy(a, null).finalize();
};

var Clone = function(a) {
  var self = this;
  return Copy(a.factory(), a);
};

var Union = function(factory, ...parts) {
  var self = this, copier, result;
  copier = CopyInto.new(factory);
  result = null;
  parts.each(function(part) {
    return result = copier.copy(part, result);
  });
  return result.finalize();
};

var union = function(a, b) {
  var self = this, f;
  f = Factory.SchemaFactory.new(a.graph_identity().schema());
  return Union(f, a, b);
};

class CopyInto {
  static new(...args) { return new CopyInto(...args) };

  constructor(factory) {
    var self = this;
    self.memo$ = Enso.EMap.new();
    self.factory$ = factory;
  };

  copy(a, b) {
    var self = this;
    self.build(a, b);
    return self.link(true, a, b);
  };

  build(a, b) {
    var self = this, new_V, a_val, b_val;
    if (! (a == null)) {
      if ((a && b) && a.schema_class().name() != b.schema_class().name()) {
        self.raise(Enso.S("Union of incompatible objects ", a, " and ", b));
      }
      self.memo$ .set$(a.identity(), new_V = b || self.factory$.get$(a.schema_class().name()));
      return new_V.schema_class().fields().each(function(field) {
        a_val = ((function() {
          try {
            return a.get$(field.name());
          } catch (DUMMY) {
            return null;
          }
        })());
        b_val = ((function() {
          try {
            return b.get$(field.name());
          } catch (DUMMY) {
            return null;
          }
        })());
        if (! (a_val == null) || ! (b_val == null)) {
          if (Enso.System.test_type(field.type(), "Primitive")) {
            if (! (a_val == null)) {
              if ((a && b) && a_val != b_val) {
                puts(Enso.S("UNION WARNING: changing ", new_V, ".", field.name(), " from '", b_val, "' to '", a_val, "'"));
              }
              return new_V .set$(field.name(), a_val);
            }
          } else if (field.traversal()) {
            if (! field.many()) {
              return self.build(a_val, b_val);
            } else {
              return a_val.each_with_match(function(a_item, b_item) {
                return self.build(a_item, b_item);
              }, b_val);
            }
          }
        } else if (! (new_V.get$(field.name()) == null)) {
          return puts(Enso.S("skipping ", new_V, ".", field.name(), " as ", new_V.get$(field.name())));
        }
      });
    }
  };

  link(traversal, a, b) {
    var self = this, new_V, a_val, b_val, val, item;
    if (a == null) {
      return b;
    } else {
      new_V = self.memo$.get$(a.identity());
      if (! new_V) {
        self.p(self.memo$);
        self.raise(Enso.S("Traversal did not visit every object a=", a, " b=", b));
      }
      if (traversal) {
        a.schema_class().fields().each(function(field) {
          a_val = a.get$(field.name());
          b_val = b && b.get$(field.name());
          if (! Enso.System.test_type(field.type(), "Primitive")) {
            if (! field.many()) {
              val = self.link(field.traversal(), a_val, b_val);
              return new_V .set$(field.name(), val);
            } else {
              return a_val.each_with_match(function(a_item, b_item) {
                item = self.link(field.traversal(), a_item, b_item);
                return new_V.get$(field.name()).push(item);
              }, b_val);
            }
          }
        });
      }
      return new_V;
    }
  };
};

Union = {
  Copy: Copy,
  Clone: Clone,
  Union: Union,
  union: union,
  CopyInto: CopyInto,
};
module.exports = Union ;
